emukit/bayesian_optimization/acquisitions/constrained_bayesian_optimization.py

```diff
@@ -0,0 +1,68 @@
+# Copyright 2018 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+
+import numpy as np
+
+from ...bayesian_optimization.acquisitions import ExpectedImprovement, ProbabilityOfFeasibility
+from ...core.acquisition import Acquisition
+from ...core.interfaces import IModel
+from ...core.loop import FixedIntervalUpdater, OuterLoop, SequentialPointCalculator
+from ...core.loop.loop_state import create_loop_state
+from ...core.optimization import AcquisitionOptimizer
+from ...core.parameter_space import ParameterSpace
+
+
+class ConstrainedBayesianOptimizationLoop(OuterLoop):
+
+    def __init__(self, space: ParameterSpace, 
+                 model_objective: IModel, 
```

Comment:
ExpectedImprovement needs IDifferentiable, so this should be `Union[IModel, IDifferentiable]`

Reply:
what about `model_constraint` ?
do we need its gradients when optimising the acquisition that is the product of 
the two acquisitions?

Reply:
I am just following the interfaces here. `model_constraint` goes into the new acquisition which delcares that it needs gradients. Maybe it in fact does not need gradients, in which case you should remove them from that acquisition